addons/account/wizard/account_transfer_wizard.py

```diff
@@ -0,0 +1,225 @@
+        if self._context.get('active_model') == 'account.move.line':
+            aml = self.env['account.move.line'].browse(self._context.get('active_ids', []))
+
+            company = aml.mapped('company_id')
+            if len(company) > 1:
+                raise UserError(_("You cannot create a transfer for entries belonging to different companies."))
+
+            for record in self:
+                record.move_line_ids = aml
+                record.company_id = company
+        else:
+            raise UserError(_("Transfer wizard should only be called on account.move.line objects"))
+
+    @api.depends('destination_account_id')
+    def _compute_display_currency_helper(self):
+        for record in self:
+            record.display_currency_helper = bool(record.destination_account_id.currency_id)
+
+    @api.depends('destination_account_id')
+    def _compute_aml_preview_data(self):
+        for record in self:
+            record.aml_preview_data = json.dumps(self._get_lines_to_create_dict())
+
+    def _get_lines_to_create_dict(self):
+        lines_data = self._get_lines_data()
+
+        line_vals = []
+        for partner in lines_data['all_partners']:
+            vals_by_currency = {}
+
+            for currency in lines_data['all_currencies']:
+                currency_partner_lines = lines_data['lines_to_transfer'].filtered(lambda x: x.partner_id == partner and x.currency_id == currency)
+                counterpart_balance = sum(currency_partner_lines.mapped('balance'))
+
+                if currency:
+                    counterpart_amount_currency = sum(currency_partner_lines.mapped('amount_currency'))
+                    counterpart_currency = currency
+                else:
+                    counterpart_amount_currency = 0
+                    counterpart_currency = self.env['res.currency']
+
+                if self.destination_account_id.currency_id:
+                    counterpart_amount_currency = self.company_id.currency_id._convert(counterpart_balance, self.destination_account_id.currency_id, self.company_id, self.date)
+                    counterpart_currency = self.destination_account_id.currency_id
+
+                if not self.company_id.currency_id.is_zero(counterpart_balance):
+                    # We store everything in a temporary dict to manage the case where
+                    # multiple lines in foreign currency have to be transferred to an
+                    # account using a currency_id, so that only one line is created,
+                    # using the currency of the account.
+                    if not vals_by_currency.get(counterpart_currency):
+                        vals_by_currency[counterpart_currency] = {'balance': 0, 'amount_currency': 0}
+
+                    vals_by_currency[counterpart_currency]['balance'] +=  counterpart_balance
+                    vals_by_currency[counterpart_currency]['amount_currency'] += counterpart_amount_currency
```

Comment:
I'm probably missing something but I don't understand why this is so complicated. Why not:
```
groupby_data = {}

target_currency = self.destination_account_id.currency_id
target_currency = None if target_currency == self.company_id.currency_id else target_currency
for line in lines_data['lines_to_transfer']:

    # Deduct currency.
    if target_currency:
        amount_currency = line.company_currency_id._convert(line.balance, target_currency, line.company_id, self.date)
        balance = line.balance
        currency = target_currency
    elif line.currency_id:
        amount_currency = line.amount_currency
        balance = line.currency_id._convert(line.amount_currency, line.company_currency_id, line.company_id, self.date)
        currency = line.currency_id
    else:
        amount_currency = 0.0
        currency = False

    # Groupby key.
    key = (line.partner, currency)
    groupby_data.setdefault(key, {'balance': 0.0, 'amount_currency': 0.0})
    groupby_data[key] += line.balance
    groupby_data[key] += amount_currency

...
```